### PR TITLE
[6.1.0]Fix symlink file creation overhead

### PR DESCRIPTION
**Pull request by kshyanashree**
The cost of symlink action scales with the size of input because Bazel re-calculates the digest of the output by following the symlink in `actuallyCompleteAction` (#14125). However, the re-calculation is redundant because the digest was already computed by Bazel when checking the outputs of the generating action. Bazel should be smart enough to reuse the result.

There is a global cache in Bazel for digest computation. Symlink action didn't make use of the cache because it uses the path of symlink as key to look up the cache. This PR changes to use the path of input file (i.e. target path) to query the cache to avoid recalculation.

For a large target (700MB), the time for symlink action is reduced from 2000ms to 1ms.

Closes #17478.

PiperOrigin-RevId: 509524641
Change-Id: Id3c9dc07d68758770c092f6307e2433dad40ba10

### CODE REVIEW

**Comment by keertk** [comment]
@coeuvre could you please approve and merge? Fyi - added `OutputPermissions.java` which was not part of PR https://github.com/bazelbuild/bazel/pull/17478.

File: src/main/java/com/google/devtools/build/lib/vfs/OutputPermissions.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
+// Copyright 2022 The Bazel Authors. All rights reserved.
```
Comment (coeuvre) [inline src/main/java/com/google/devtools/build/lib/vfs/OutputPermissions.java:1]: Please remove this file. It's not relevant with this PR.

File: src/test/java/com/google/devtools/build/lib/skyframe/ActionMetadataHandlerTest.java [hunk 2]
```diff
@@ -41,6 +41,8 @@
 import com.google.devtools.build.lib.testutil.Scratch;
 import com.google.devtools.build.lib.util.io.TimestampGranularityMonitor;
 import com.google.devtools.build.lib.vfs.DigestHashFunction;
+import com.google.devtools.build.lib.vfs.DigestUtils;
+import com.google.devtools.build.lib.vfs.OutputPermissions;
```
Comment (coeuvre) [inline src/test/java/com/google/devtools/build/lib/skyframe/ActionMetadataHandlerTest.java:45]: Ditto: please remove this line.

**Review by coeuvre — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/devtools/build/lib/vfs/OutputPermissions.java (hunk 1 above)
Comment (keertk) [reply src/main/java/com/google/devtools/build/lib/vfs/OutputPermissions.java]: Done

File: src/test/java/com/google/devtools/build/lib/skyframe/ActionMetadataHandlerTest.java (hunk 2 above)
Comment (keertk) [reply src/test/java/com/google/devtools/build/lib/skyframe/ActionMetadataHandlerTest.java]: Done

**Review by coeuvre — APPROVED** [review]

**Pull request merged**